File: systems-engineering/performance-engineering/impl/profiling_lab/python/cpu_hog.py
#!/usr/bin/env python3
"""Simple CPU-bound HTTP server for profiling demo."""

import http.server
import socketserver
import time


def fib(n: int) -> int:
    if n <= 1:
        return n
    return fib(n - 1) + fib(n - 2)


class Handler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/work":
            start = time.perf_counter()
            result = fib(38)
            duration = time.perf_counter() - start
            body = f"result={result} duration={duration:.3f}s\n".encode()
            self.send_response(200)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(body)
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        pass  # suppress default logging


if __name__ == "__main__":
    PORT = 8080
    print(f"Listening on :{PORT}")
    print("Profile workflow:")
    print("  1. python -m cProfile -o stats.prof cpu_hog.py")
    print("  2. curl http://localhost:8080/work")
    print("  3. Ctrl-C to stop server")
    print("  4. flameprof stats.prof > flame.svg")
    with socketserver.TCPServer(("", PORT), Handler) as httpd:
        httpd.serve_forever()

File: systems-engineering/performance-engineering/impl/profiling_lab/js/cpu_hog.js
import http from "http";

function fib(n) {
  if (n <= 1) return n;
  return fib(n - 1) + fib(n - 2);
}

const server = http.createServer((req, res) => {
  if (req.url === "/work") {
    const start = Date.now();
    const result = fib(38);
    res.writeHead(200, { "Content-Type": "text/plain" });
    res.end(`result=${result} duration=${Date.now() - start}ms\n`);
  } else {
    res.writeHead(404);
    res.end();
  }
});

server.listen(8080, () => {
  console.log("Listening on :8080");
  console.log("Profile workflow:");
  console.log("  1. node --prof cpu_hog.js   (or npx tsx cpu_hog.ts)");
  console.log("  2. curl http://localhost:8080/work");
  console.log("  3. node --prof-process isolate-*.log > profile.txt");
  console.log("Or use 0x: npx 0x cpu_hog.js");
});
